src/vm/types/serialization.rs

```diff
@@ -0,0 +1,881 @@
+                }
+                let tuple_type = match expected_type {
+                    None => None,
+                    Some(TypeSignature::Atom(AtomTypeIdentifier::TupleType(tuple_type))) => Some(tuple_type),
+                    Some(x) => return Err(InterpreterError::DeserializeExpected(x.clone()).into())
+                };
+
+                let deserialized_entries: InterpreterResult<_> = entries
+                    .drain()
+                    .map(|(key, json_val)| {
+                        let expected_field_type = match tuple_type {
+                            None => None,
+                            Some(some_tuple) => Some(
+                                some_tuple
+                                    .field_type(&key)
+                                    .ok_or_else(|| RuntimeErrorType::ParseError(
+                                        format!("Expected tuple type does not contain field '{}' but JSON does.", key)))?)
+                        };
+                        Ok((key, Value::try_deserialize_parsed(json_val, expected_field_type)?))
+                    })
+                    .collect();
+                if let Some(tuple_type) = tuple_type {
+                    TupleData::from_data_typed(deserialized_entries?, tuple_type)
+                        .map(Value::from)
+                } else {
+                    TupleData::from_data(deserialized_entries?)
+                        .map(Value::from)
+                }
+            },
+            JSONParser::ContractPrincipal { type_n, issuer, name } => {
+                if type_n != TYPE_CONTRACT_PRINCIPAL {
+                    return Err(InterpreterError::DeserializeUnexpectedTypeField(type_n).into())
+                }
+                check_match!(expected_type, AtomTypeIdentifier::PrincipalType)?;
+                Ok(Value::from(
+                    if issuer == ":none:" {
+                        PrincipalData::ContractPrincipal(name)
+                    } else {
+                        PrincipalData::QualifiedContractPrincipal {
+                            sender: PrincipalData::parse_standard_principal(&issuer)?,
+                            name }
+                    }
+                ))
+            }
+        }
+    }
+
+    pub fn try_deserialize(json: &str, expected: &TypeSignature) -> InterpreterResult<Value> {
+        let data: JSONParser = serde_json::from_str(json)?;
+        Value::try_deserialize_parsed(data, Some(expected))
+    }
+
+    pub fn try_deserialize_untyped(json: &str) -> InterpreterResult<Value> {
+        let data: JSONParser = serde_json::from_str(json)?;
+        Value::try_deserialize_parsed(data, None)
+    }
+}
+
+impl Value {
+    pub fn deserialize(json: &str, expected: &TypeSignature) -> Self {
+        Value::try_deserialize(json, expected)
+            .expect("ERROR: Failed to parse Clarity JSON")
+    }
+}
+
+
+#[cfg(test)]
```

Comment:
I'm not sure how reasonable or difficult this is, so feel free to dismiss.

Would it be desirable to add an "exhaustive" test that tries to serialize/deserialize all possible type combinations up to a certain (small) recursion depth?  Might be the job of a fuzzer to do this; just something to think about.

Reply:
Yeah -- the Value enum is fairly amenable to such a test.